docs/features/nullable-metadata.md

```diff
@@ -0,0 +1,125 @@
+Nullable Metadata
+=========
+The following describes the representation of nullable annotations in metadata.
+
+### NullableAttribute
+Type references are annotated in metadata with a `NullableAttribute`.
+
+```C#
+namespace System.Runtime.CompilerServices
+{
+    [AttributeUsage(
+        AttributeTargets.Class |
+        AttributeTargets.Event |
+        AttributeTargets.Field |
+        AttributeTargets.GenericParameter |
+        AttributeTargets.Parameter |
+        AttributeTargets.Property |
+        AttributeTargets.ReturnValue,
+        AllowMultiple = false)]
+    public sealed class NullableAttribute : Attribute
+    {
+        public readonly byte[] NullableFlags;
+        public NullableAttribute(byte b)
+        {
+            NullableFlags = new byte[] { b };
+        }
+        public NullableAttribute(byte[] b)
+        {
+            NullableFlags = b;
+        }
+    }
+}
+```
+
+Each type reference is accompanied by a `NullableAttribute` with an array of bytes, where 0 is oblivious, 1 is not annotated and 2 is annotated.
+
+Value types may be marked with 0 (oblivious) or 1 (not annotated) - the representations are considered equivalent.
+
+To optimize trivial cases the attribute may be omitted or replaced with an attribute that takes a single byte value rather than an array. The C#8 compiler uses the following optimizations.
+
+Optimized cases:
+1. All parts are oblivious and the [context](#NullableContextAttribute) is oblivious: the `NullableAttribute` is omitted
+2. All parts are oblivious and the [context](#NullableContextAttribute) is not annotated: `NullableAttribute(0)` (rather than `new byte[] { 0 }`)
+3. All parts are not annotated and the [context](#NullableContextAttribute) is not annotated: the `NullableAttribute` is omitted
+4. All parts are not annotated and the [context](#NullableContextAttribute) is oblivious: `NullableAttribute(1)` (rather than `new byte[] { 1 }`)
+5. All parts are annotated: `NullableAttribute(2)` (rather than `new byte[] { 2 }`)
+
+The first case, where all parts are oblivious in an oblivious context, matches the representation in legacy assemblies where there are no `NullaleAttribute` annotations.
+
+`NullableAttribute(1)` should be placed on a type parameter definition that has a `class!` constraint.
```

Comment:
Are attributes in this list subject to ```NullableContextAttribute``` optimizations? If yes, we might need to emit ```NullableAttribute(0)``` within some contexts